core/src/main/java/io/grpc/internal/MessageSender.java

```diff
@@ -0,0 +1,293 @@
+  private Phase phase = Phase.HEADERS;
+  private boolean headersSent;
+  private Metadata stashedTrailers;
+
+  /**
+   * The number of bytes currently queued, waiting to be sent. When this falls below
+   * onReadyThreshold, {@link StreamListener#onReady()} will be called.
+   */
+  @GuardedBy("onReadyLock")
+  private int numSentBytesQueued;
+
+  private final Object onReadyLock = new Object();
+
+  MessageSender(WritableBufferAllocator bufferAllocator, Handler handler, boolean server) {
+    this.handler = checkNotNull(handler, "handler");
+    this.server = server;
+    MessageFramer.Sink frameHandler = new MessageFramer.Sink() {
+      @Override
+      public void deliverFrame(WritableBuffer frame, boolean endOfStream, boolean flush) {
+        onMessageFrame(frame, endOfStream, flush);
+      }
+    };
+    framer = new MessageFramer(frameHandler, bufferAllocator);
+  }
+
+  void setCompressor(Compressor c) {
+    // TODO(carl-mastrangelo): check that headers haven't already been sent.  I can't find where
+    // the client stream changes outbound phase correctly, so I am ignoring it.
+    framer.setCompressor(c);
+  }
+
+  /**
+   * Request the sending of headers. Must be called from the channel thread context.
+   */
+  void sendHeaders(Metadata headers) {
+    checkNotNull(headers, "headers");
+    phase(Phase.HEADERS);
+    headersSent = true;
+    handler.sendHeaders(headers);
+    phase(Phase.MESSAGE);
+  }
+
+  /**
+   * Request that a message be sent. Must be called from the channel thread context.
+   */
+  void sendMessage(InputStream message) {
+    checkNotNull(message);
+    phase(Phase.MESSAGE);
+    if (!framer.isClosed()) {
+      framer.writePayload(message);
+    }
+  }
+
+  /**
+   * Request to flush any buffered messages. Must be called form the channel thread context.
+   */
+  void flush() {
+    if (!framer.isClosed()) {
+      framer.flush();
+    }
+  }
+
+  /**
+   * If {@code true}, indicates that the transport is capable of sending additional messages without
+   * requiring excessive buffering internally. Otherwise, {@link Handler#onReady()} will be
+   * called when it turns {@code true}.
+   *
+   * <p>This is just a suggestion and the application is free to ignore it, however doing so may
+   * result in excessive buffering within the transport.
+   */
+  boolean isReady() {
+    if (phase != Phase.STATUS) {
+      synchronized (onReadyLock) {
+        return numSentBytesQueued < ONREADY_THRESHOLD;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Closes this outbound context. Must be called from the application thread context.
+   */
+  void close(Status status, boolean send) {
+    checkNotNull(status, "status");
+    if (phase(Phase.STATUS) == Phase.STATUS) {
+      // Already closed.
+      return;
+    }
+
+    if (!send) {
+      // Close the framer, but do not flush any remaining data.
+      framer.dispose();
+      return;
+    }
+
+    if (server) {
+      // TODO(louiscryan): Remove
```

Comment:
nit: TODO's should have a when or a why.  That way one of us can look at this in a few months and know if it can be removed.